proxy/source/markdown_parser.py

```diff
@@ -58,22 +54,36 @@ def _convert_crlf(input_str: str) -> str:
 
 
 def _parse_links(input_str: str) -> str:
+    input_str = re.sub(
+        r"\[(.+?)\]\((https?:\/\/[-a-zA-Z0-9._~:/?#@!$&()*+,;=%']+)\)",
+        r'<a href="\2" target="_blank" rel="nofollow noreferrer noopener">\1</a>',
+        input_str,
+        flags=re.MULTILINE|re.IGNORECASE,
+    )
     return re.sub(
-        r"\[([\w\W]+?)\]\(([\w\W]+?)\)",
-        r'<a href="\2">\1</a>',
+        r"(?<!href=\")(https?:\/\/[-a-zA-Z0-9._~:/?#@!$&()*+,;=%']+)",
+        r'<a href="\1" target="_blank" rel="nofollow noreferrer noopener">\1</a>',
         input_str,
-        flags=re.MULTILINE,
+        flags=re.MULTILINE|re.IGNORECASE,
     )
 
 
 def _parse_headers(input_str: str) -> str:
-    return re.sub(r"^#+ +([\w\W]+?)\n", r"<h3>\1</h3>\n", input_str, flags=re.MULTILINE)
+    search = re.finditer(r"^(#{1,5}) +(.+)[# ]?$", input_str, re.MULTILINE)
+    for i in search:
+        h = 2 - (min(len(i.group(1)) + 1, 5) / 10)
+        input_str = re.sub(
+            i.group(),
+            f'<span style="font-size: {h}em;">{i.group(2)}</span>',
+            input_str,
+        )
+    return input_str
 
 
 def _parse_strong_emphasis(input_str: str) -> str:
     return "\n".join(
         re.sub(
-            r"(?:\*\*|\_\_)([\w]+?[\w\W]+?[\w]+?)(?:\*\*|\_\_)",
+            r"(?<!\\)\*\*(\w.*?)(?<!\\)\*\*|(?<!\\)__(\w.*?)(?<!\\)__",
```

Comment:
Nice, I'm a fan of also handling the escaped character.

I think you're handling the first character being a space, but not the last?

```python
>>> text = """
... **asdf**
... ** asdf**
... **asdf **
... ** asdf **
... """
>>> print(_parse_strong_emphasis(text))

<strong>asdf</strong>
** asdf**
<strong>asdf </strong>
** asdf **
```

It's fine though, I'm not going to block on this (especially since the original impl was flawed as it only allowed `\w` which ignores symbols as valid characters).